components/tidb_query/src/rpn_expr/impl_math.rs

```diff
@@ -463,6 +463,46 @@ pub fn round_dec(arg: &Option<Decimal>) -> Result<Option<Decimal>> {
     }
 }
 
+#[inline]
+#[rpn_fn]
+pub fn truncate_real_with_int(arg0: &Option<Real>, arg1: &Option<Int>) -> Result<Option<Real>> {
+    match (arg0, arg1) {
+        (&Some(x), &Some(d)) => {
+            let d = if d >= 0 {
+                d.min(i64::from(i32::max_value())) as i32
+            } else {
+                d.max(i64::from(i32::min_value())) as i32
+            };
+            Ok(Some(truncate_real(x, d)))
+        }
+        _ => Ok(None),
+    }
+}
+
+#[inline]
+#[rpn_fn]
+pub fn truncate_real_with_uint(arg0: &Option<Real>, arg1: &Option<Int>) -> Result<Option<Real>> {
+    match (arg0, arg1) {
+        (&Some(x), &Some(d)) => {
+            let d = (d as u64).min(i32::max_value() as u64) as i32;
+            Ok(Some(truncate_real(x, d)))
+        }
+        _ => Ok(None),
+    }
+}
+
+pub fn truncate_real(x: Real, d: i32) -> Real {
+    let shift = 10_f64.powi(d);
+    let tmp = x * shift;
+    if *tmp == 0_f64 {
```

Comment:
Why do we need this zero case, is it repeatedly covered with `Real::from(tmp.trunc() / shift)`? BTW, could we rename `tmp` as `result`?

Reply:
Maybe `shift` will be 0 when `d` is a very small negative number, such as `truncate(123.1234, -6666666)`. Therefore, `tmp.trunc() / shift` will be `NaN`.

Reply:
@iosmanthus I referred `truncate_real` in TiKV and TiDB, these functions name `x * shift` with `tmp`. Maybe it is because `x * shift` is a temporary intermediate result?

Reply:
@iosmanthus I referred `truncate_real` in TiKV and TiDB, these functions name `x * shift` with `tmp`. Maybe it is because `x * shift` is a temporary intermediate result?